compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13314,113 +13577,92 @@ OMR::Z::TreeEvaluator::primitiveArraycopyEvaluator(TR::Node* node, TR::CodeGener
-
-      // Generate OOL Slow Path
-      TR_S390OutOfLineCodeSection* outOfLineCodeSection = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(slowPathLabel, mergeLabel, cg);
-      cg->getS390OutOfLineCodeSectionList().push_front(outOfLineCodeSection);
-      outOfLineCodeSection->swapInstructionListsWithCompilation();
-
-      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, slowPathLabel);
-
-      TR::Register* strideRegister = cg->allocateRegister();
-
-      MemCpyVarLenTypedMacroOp vtOpSlow(node, byteDstNode, byteSrcNode, cg, node->getArrayCopyElementType(), byteLenReg, byteLenNode, true, node->isForwardArrayCopy());
-      vtOpSlow.generate(byteDstReg, byteSrcReg, strideRegister);
-      vtOpSlow.cleanUpReg();
-
-      cg->stopUsingRegister(strideRegister);
-
-      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergeLabel);
-
-      // Switch back from OOL Slow Path
-      outOfLineCodeSection->swapInstructionListsWithCompilation();
+      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, startICFLabel);
+      startICFLabel->setStartInternalControlFlow();
+      cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpOpCode(), node, byteLenReg, static_cast<int8_t>(0x0), TR::InstOpCode::COND_BNH, mergeLabel, false);
+      iComment("byteLen <= 0 goto mergeLabel");
       }
-#endif
 
    if (node->isForwardArrayCopy())
       {
-      if (byteLenNode->getOpCode().isLoadConst())
-         {
-         MemCpyConstLenMacroOp op(node, byteDstNode, byteSrcNode, cg, byteLenNode->getConst<int64_t>());
-
-         op.generate(byteDstReg, byteSrcReg);
-         }
-      else
-         {
-         bool alreadyEvaluatedLengthMinusOne = false;
-
-         // Don't evaluate length minus one if byteLenNode was already evaluated
-         if (!alreadyEvaluatedByteLenNode)
-            {
-            byteLenReg = cg->evaluateLengthMinusOneForMemoryOps(byteLenNode, true, alreadyEvaluatedLengthMinusOne);
-            }
-
-         MemCpyVarLenMacroOp op(node, byteDstNode, byteSrcNode, cg, byteLenReg, byteLenNode, alreadyEvaluatedLengthMinusOne);
-
-         op.setUseEXForRemainder(true);
-         op.generate(byteDstReg, byteSrcReg);
-         }
+      // Optimizer has already identified that this is array copy in forward direction.
+      TR::TreeEvaluator::forwardArrayCopySequenceGenerator(node, cg, byteSrcReg, byteDstReg, byteLenReg, byteLenNode, srm, mergeLabel);
       }
    else
       {
-      // Don't clobber evaluate byteLenNode again if it was already evaluated
-      if (!alreadyEvaluatedByteLenNode)
+      // We need to decide direction of array copy at runtime.
+      TR::LabelSymbol *forwardArrayCopyLabel = generateLabelSymbol(cg);
+      cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpRegOpCode(), node, byteSrcReg, byteDstReg, TR::InstOpCode::COND_BNL, forwardArrayCopyLabel, false);
```

Comment:
See PoPs 7-44 where `BPP` instruction is documented. Check the Programming Notes about assumed branch-prediction logic. I'm not sure what instruction ends up being generated here and for the branch below but as I see the most common scenario will be a forward array copy so one of the branches you generate will always (more or less) be taken and one will always fail. Let's not introduce unwanted branch mispredictions and make sure the branches we're generating here are the optimal ones.

i.e. my question is should this be a `TR::InstOpCode::COND_BL` branch?

Reply:
The sequence favors executing arrayCopy Sequence. I generated here are like this.
```
if (srcPointer >= dstPointer)
   goto forwardSequenceLabel
else if (srcPointer+byteLength > dstPointer)
   goto backwardSequenceLabel

forwardSequenceLabel: 
...
Jump mergeLabel
backwardSequenceLabel:
...
mergeLabel:
```
So in summary, it first checks if srcPointer is greater or equal to dstPointer and if that is case then it can do arrayCopy in forward direction even if there is an overlap and that is what I check first and in this case I won't even need to know if overlap exists or not.
If srcPointer is less than dstPointer then I need to check if there is overlap and in case of overlap does arraycopy in backward direction.